src/ppx-metaquot/fun_ty.mli

```diff
@@ -0,0 +1,143 @@
+
+    This module also serves as a runtime library for the customized
+    [ppx_metaquot], so that one may write:
+
+    [let p = [%funty: int -> bool]], and get:
+
+    [val p : ((int -> bool) Ty.ty, int -> unit, bool) fun_ty = <abstr>] *)
+
+(** GADT for arguments of n-ary functions, implemented as nonempty
+    heterogeneous lists.
+
+    Usage: [arg 3 @@ arg "word" @@ last false]
+
+    Alternatively: [3 @: "word" @:!! false] *)
+type ('arrow, 'uarrow, 'ret) args
+
+(** [last e], or equivalently [!! e], builds a one-element argument list *)
+val last :
+  'a ->
+  ('a -> 'ret, 'a -> unit, 'ret) args
+
+(** [arg a l], or equivalently [a @: l], adds [a] in front of the
+   argument list [l] *)
+val arg :
+  'a ->
+  ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  ('a -> 'ar -> 'row, 'a -> 'ar -> 'urow, 'ret) args
+
+(** Helper notation for [last] *)
+val (!!) :
+  'a ->
+  ('a -> 'ret, 'a -> unit, 'ret) args
+
+(** Helper notation for [arg] *)
+val (@:) :
+  'a ->
+  ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  ('a -> 'ar -> 'row, 'a -> 'ar -> 'urow, 'ret) args
+
+(** [a @:!! l] is another notation for [a @: !! l] (with a space) *)
+val (@:!!) :
+  'a -> 'b ->
+  ('a -> 'b -> 'ret, 'a -> 'b -> unit, 'ret) args
+
+(** [apply f l] applies a n-ary function [f] to the arguments from [l] *)
+val apply :
+  ('ar -> 'row) -> ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  'ret
+
+(** GADT for function types.
+
+    Given an arrow type ['a -> 'row], the following construct provides
+    a more precise representation of this function type than
+    [[%ty: 'a -> 'row] : ('a -> 'row) Ty.ty]:
+
+    [[%funty: 'a -> 'row] : (('a -> 'row) Ty.ty, 'a -> 'urow, 'ret) fun_ty]
+
+    In particular, the codomain type ['ret] is made explicit, so that
+    if ['row = 'b -> 'c], we get ['urow = 'b -> unit] and ['ret = 'c].
+
+    Usage: [arg_ty [%ty: int] @@ arg_ty [%ty: string] @@
+            last_ty [%ty: bool] [%ty: unit]]
+
+    Alternatively: [[%funty: int -> string -> bool -> unit]] *)
+type ('arrow, 'uarrow, 'ret) fun_ty
+
+(** [last_ty [%ty: a] [%ty: r]] builds a function type for [a -> r] *)
+val last_ty :
+  'a Ty.ty ->
+  'ret Ty.ty ->
+  (('a -> 'ret) Ty.ty, 'a -> unit, 'ret) fun_ty
+
+(** [arg_ty [%ty: a] [%funty: b ->...-> r]] builds a function type for
+    [a -> b ->...-> r] *)
+val arg_ty :
+  'a Ty.ty ->
+  (('ar -> 'row) Ty.ty, 'ar -> 'urow, 'ret) fun_ty ->
+  (('a -> 'ar -> 'row) Ty.ty, ('a -> 'ar -> 'urow), 'ret) fun_ty
+
+(** [ty_of_fun_ty funty] returns a term of type [('ar -> 'row) Ty.ty],
+    assuming [funty : (('ar -> 'row) Ty.ty, _, _) fun_ty] *)
+val ty_of_fun_ty :
+  (('ar -> 'row) Ty.ty, 'ar -> 'urow, 'ret) fun_ty ->
+  ('ar -> 'row) Ty.ty
+
+(** [get_ret_ty funty] returns a term of type ['ret Ty.ty], assuming
+   [funty : (_ , _, 'ret) fun_ty] *)
+val get_ret_ty :
+  ('p -> 'a) Ty.ty -> ('p -> 'a, 'p -> 'c, 'ret) args -> 'ret Ty.ty
+
+
+(** Signature [S] is intended to be instantiated in [Test_lib] with:
+    [module M = struct
```

Comment:
@yurug you're right, `signature … is intended to be instantiated` was not properly formulated. (Because we usually instantiate a functor, not a signature.) 

The main goal of this ocamldoc comment was just to prevent the signature from being undocumented, while mentioning an example of use of the functor `Make` that depends on this signature… But indeed, it was probably not that relevant.

So what would be the better documentation choice here, in your opinion?

1. Replace `Signature [S] is intended to be instantiated in [Test_lib] with: [module M = …].` with
   `Signature [S] is useful to define the functor [Make] below, used in [Test_lib].`
2. Remove that ocamldoc comment and leave the signature with no ocamldoc comment.

Kind regards, Erik

Reply:
Hi @yurug, FYI I've pushed just now a new version of the doc (following the "choice 2." for `S`) after polishing some other ocamldoc comments (adding e.g. some missing dots).
Does this version look good to you?